phygnn/layers/custom_layers.py

```diff
@@ -943,6 +941,89 @@ def call(self, x):
         return output
 
 
+class AxialAttentionBlock(tf.keras.layers.Layer):
+    """Axial Self-Attention block
+
+    References
+    ----------
+    1. Axial Attention in Multidimensional Transformers.
+       https://arxiv.org/abs/1912.12180
+    """
+
+    def __init__(self, num_heads=4, key_dim=64, name=None):
+        """
+        Parameters
+        ----------
+        num_heads : int
+            Number of attention heads
+        key_dim : int
+            Size of each attention head
+        name : str | None
+            Name of layer.
+        """
+
+        super().__init__(name=name)
+        self.num_heads = num_heads
+        self.key_dim = key_dim
+        self.attention_layers = None
+        self.rank = None
+
+    def build(self, input_shape):
+        """Build the AxialAttentionBlock layer based on an input shape
+
+        Parameters
+        ----------
+        input_shape : tuple
+            Shape tuple of the input tensor
+        """
+        self.rank = len(input_shape)
+        self.attention_layers = [
+            tf.keras.layers.MultiHeadAttention(
+                num_heads=self.num_heads, key_dim=self.key_dim
+            )
+            for _ in range(self.rank - 2)
```

Comment:
-2 because no attention in obs or feature axis, right? Recommend adding `assert self.rank >= 3`